--- common/lib/modules/fglrx/build_mod/firegl_public.h.orig	2016-08-01 10:29:07.635774314 +0200
+++ common/lib/modules/fglrx/build_mod/firegl_public.h	2016-08-01 10:31:39.945556616 +0200
@@ -647,12 +647,20 @@
 #endif
 
 #ifndef cpu_has_pat
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(4,7,0)
+#define cpu_has_pat boot_cpu_has(X86_FEATURE_PAT)
+#else
 #define cpu_has_pat  test_bit(X86_FEATURE_PAT, (void *) &boot_cpu_data.x86_capability)
 #endif
+#endif
 
 #ifndef cpu_has_pge
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(4,7,0)
+#define cpu_has_pge boot_cpu_has(X86_FEATURE_PGE)
+#else
 #define cpu_has_pge test_bit(X86_FEATURE_PGE, &boot_cpu_data.x86_capability)
 #endif
+#endif
 
 /* 2.6.29 defines pgprot_writecombine as a macro which resolves to a
  * GPL-only function with the same name. So we always use our own
