--- a/Makefile	2018-04-25 18:54:11.618745948 -0700
+++ b/Makefile	2018-04-25 19:15:32.849809029 -0700
@@ -44,21 +44,19 @@
 
 UGID_USE_RWLOCK = 0
 
-OPTS 	    = -O2
 SRC	    = $(wildcard src/*.cpp)
 OBJ         = $(SRC:src/%.cpp=obj/%.o)
 DEPS        = $(OBJ:obj/%.o=obj/%.d)
 TARGET      = mergerfs
 MANPAGE     = $(TARGET).1
-FUSE_CFLAGS = -D_FILE_OFFSET_BITS=64 -Ilibfuse/include
-CFLAGS      = -g -Wall \
-	      $(OPTS) \
+FUSE_CFLAGS = -D_FILE_OFFSET_BITS=64
+MY_CFLAGS      = -Wall \
 	      -Wno-unused-result \
               $(FUSE_CFLAGS) \
               -DFUSE_USE_VERSION=29 \
               -MMD \
 	      -DUGID_USE_RWLOCK=$(UGID_USE_RWLOCK)
-LDFLAGS       = -pthread -lrt
+MY_LDFLAGS       = -pthread -lrt
 
 PREFIX        = /usr/local
 EXEC_PREFIX   = $(PREFIX)
@@ -84,9 +82,8 @@
 	@echo "usage: make"
 	@echo "make XATTR_AVAILABLE=0 - to build program without xattrs functionality (auto discovered otherwise)"
 
-$(TARGET): version obj/obj-stamp libfuse/lib/.libs/libfuse.a $(OBJ)
-	cd libfuse && make
-	$(CXX) $(CFLAGS) $(OBJ) -o $@ libfuse/lib/.libs/libfuse.a -ldl $(LDFLAGS)
+$(TARGET): version obj/obj-stamp $(OBJ)
+	$(CXX) $(CXXFLAGS) $(MY_CFLAGS) $(OBJ) -o $@ -ldl -lfuse $(LDFLAGS) $(MY_LDFLAGS)
 
 mount.mergerfs: $(TARGET)
 	$(LN) -fs "$<" "$@"
@@ -109,17 +106,16 @@
 	$(TOUCH) $@
 
 obj/%.o: src/%.cpp
-	$(CXX) $(CFLAGS) -c $< -o $@
+	$(CXX) $(CXXFLAGS) $(MY_CFLAGS) -c $< -o $@
 
-clean: rpm-clean libfuse_Makefile
+clean: rpm-clean
 	$(RM) -f src/version.hpp
 	$(RM) -rf obj
 	$(RM) -f "$(TARGET)" mount.mergerfs
 	$(FIND) . -name "*~" -delete
-	cd libfuse && $(MAKE) clean
 
-distclean: clean libfuse_Makefile
-	cd libfuse && $(MAKE) distclean
+distclean: clean
+	true
 ifeq ($(GIT_REPO),1)
 	$(GIT) clean -fd
 endif
@@ -208,19 +204,4 @@
 	yum -y install git rpm-build libattr-devel gcc-c++ make which python automake libtool gettext-devel
 endif
 
-unexport CFLAGS LDFLAGS
-.PHONY: libfuse_Makefile
-libfuse_Makefile:
-ifeq ($(shell test -e libfuse/Makefile; echo $$?),1)
-	cd libfuse && \
-	$(MKDIR) -p m4 && \
-	autoreconf --force --install && \
-        ./configure --enable-lib --disable-util --disable-example
-endif
-
-libfuse/lib/.libs/libfuse.a: libfuse_Makefile
-	cd libfuse && $(MAKE)
-
-.PHONY: all clean install help version
-
 -include $(DEPS)
