From c7111e29d9c56b5a22510607cb2487e8d5282557 Mon Sep 17 00:00:00 2001
From: David Sterba <dsterba@suse.com>
Date: Tue, 5 Dec 2017 15:45:09 +0100
Subject: [PATCH 11/12] if0 zstdio

---
 grub-core/io/zstdio.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/grub-core/io/zstdio.c b/grub-core/io/zstdio.c
index 5f49b44aa..aa6060308 100644
--- a/grub-core/io/zstdio.c
+++ b/grub-core/io/zstdio.c
@@ -30,6 +30,8 @@
 
 GRUB_MOD_LICENSE ("GPLv3+");
 
+#if 0
+
 #include <include/grub/zstd.h>
 
 #define MAX_BUF_SIZE		(16384)
@@ -238,3 +240,5 @@ GRUB_MOD_FINI (zstdio)
 {
   grub_file_filter_unregister (GRUB_FILE_FILTER_ZSTDIO);
 }
+
+#endif
-- 
2.16.1

